### DISCUSSION TITLE
SqlAlchemy+pgvector+fastapi+postgresql

### DISCUSSION THREAD

**Discussion by ShivanshMaCoder** [sqlalchemy/sqlalchemy#11529] [Usage Questions]
How to write the postgresql query which is running in psycpog2 in sqlalchemy after creating a session->

(select f.embeddings,c.content from feature f left join campaign c  on f.campaign_id  = c.campaign_id where age='18-25' or age ='0-17' and gender ='Male' order by f.embeddings <=> %s Limit 5, (product_embedding,))

**Comment by CaselIT** [comment]
Hi,

What's the particular issue with that query? If you are looking at how to render that operator `<=>` you can use `op()`. https://docs.sqlalchemy.org/en/20/core/sqlelement.html#sqlalchemy.sql.expression.Operators.op

**Reply by ShivanshMaCoder** [reply]
I am able to run this query using psycopg2 connection->

```py
def get_pg_connection():
    conn = psycopg2.connect(
        user=os.environ["DB_USER"],
        password=os.environ["DB_PASS"],
        host='*.*.*.*',
        port='5432',
        database=os.environ["DB_NAME"]
        )
    return conn

conn = get_pg_connection()
    register_vector(conn)
    cur = conn.cursor()
    cur.execute(f""" SELECT  c.content as content , f.embeddings as embeddings
                FROM feature f LEFT JOIN campaign c 
                ON f.campaign_id = c.campaign_id
                {where_clause} 
                ORDER BY f.embeddings <=> %s LIMIT 5""",(embedding_array,))
    topn_docs = cur.fetchall()
    cur.close()
    conn.close()
 ```
    I am unable to do the same thing is sqlalchemy session ->
    db: Session = Depends(get_db)
    Now using db, query I want to achieve the same thing

**Reply by CaselIT** [reply]
you can refer to the documentation on how to get a driver connection https://docs.sqlalchemy.org/en/20/core/connections.html#dbapi-connections

**Comment by ShivanshMaCoder** [comment]
But I am doing like this->
```py
from typing import Dict, List, Union
from sqlalchemy import select, or_, and_, text
import time
import json
import traceback
import uvicorn
from fastapi import FastAPI, HTTPException, Depends
from pgvector.sqlalchemy import Vector
from sqlalchemy.orm import Session, sessionmaker
from sqlalchemy.exc import OperationalError
from database import connect_tcp_socket
from data_models import FeatureTable, CampaignTable , Features , Response , Campaign
from embeddings import get_embeddings
from similarity import cosine_similarity
from where_query import create_query
import numpy as np

# Connect to the database engine
engine = connect_tcp_socket()

SessionLocal = sessionmaker(autocommit=False, autoflush=False, bind=engine)

def get_db()-> Session:
    db = SessionLocal()
    try:
        yield db
    finally:
        db.close() 
app = FastAPI()

@app.get('/')
def home():
    return "Welcome User"

@app.post('/v1/content')
def content(payload:Features, db: Session = Depends(get_db)):
    db.execute(text('CREATE EXTENSION IF NOT EXISTS vector'))
```
 Where connect_tcp_socket() returns a connection pool in def content how to perform the above?

**Reply by CaselIT** [reply]
I'm not sure what you are asking. engine would be something returned by https://docs.sqlalchemy.org/en/20/core/engines.html#sqlalchemy.create_engine
